pydarn/io/borealis/restructure_borealis.py

```diff
@@ -0,0 +1,774 @@
+            scan_start_array[rec_idx] = data_dict[k]["scan_start_marker"]
+            num_slices_array[rec_idx] = data_dict[k]["num_slices"]
+
+            beam_nums_buffer = data_dict[k]["beam_nums"]
+            beam_azms_buffer = data_dict[k]["beam_azms"]
+            # the beam_nums, beam_azms at a record may have appended zeros, 
+            # which is why a num_beams array is necessary.
+            beam_nums_array[rec_idx][0:len(beam_nums_buffer)] = beam_nums_buffer 
+            beam_azms_array[rec_idx][0:len(beam_azms_buffer)] = beam_azms_buffer
+            num_beams_array[rec_idx] = len(beam_nums_buffer)
+
+            # insert data into buffer
+            record_buffer = data_dict[k]["data"]
+            data_pos = rec_idx * data_buffer_offset
+            data_end = data_pos + len(record_buffer)
+            data_buffer[data_pos:data_end] = record_buffer
+
+            # insert sequence timestamps into buffer
+            rec_sqn_ts = data_dict[k]["sqn_timestamps"]
+            sqn_pos = rec_idx * sqn_buffer_offset
+            sqn_end = sqn_pos + data_dict[k]["num_sequences"]
+            sqn_ts_buffer[sqn_pos:sqn_end] = rec_sqn_ts
+
+            rec_noise = data_dict[k]["noise_at_freq"]
+            noise_pos = rec_idx * noise_buffer_offset
+            noise_end = noise_pos + data_dict[k]["num_sequences"]
+            noise_buffer[noise_pos:noise_end] = rec_noise
+
+        # write leftover metadata and data
+        new_data_dict["num_sequences"] = sqn_num_array
+        new_data_dict["int_time"] = int_time_array
+        new_data_dict["scan_start_marker"] = scan_start_array
+        new_data_dict["num_slices"] = num_slices_array
+        new_data_dict["beam_nums"] = beam_nums_array 
+        new_data_dict["beam_azms"] = beam_azms_array
+        new_data_dict["num_beams"] = num_beams_array
+
+        new_data_dict["data"] = data_buffer.reshape(data_shape)
+        new_data_dict["sqn_timestamps"] = sqn_ts_buffer.reshape(sqn_shape)
+        new_data_dict["noise_at_freq"] = noise_buffer.reshape(noise_shape)
+
+        new_data_dict["data_descriptors"] = data_descriptors
+        # new_data_dict["data_dimensions"] = data_dims_array
+
+        return new_data_dict
+
+    @classmethod
+    def bfiq_site_to_array(cls, filename: str, data_dict: OrderedDict) -> dict:
+        """
+        Restructuring method for bfiq data.
+
+        Parameters
+        ----------
+        filename
+            String of where file originated from for better error messages.
+        data_dict
+            a dict of timestamped records loaded from an hdf5 Borealis bfiq
+            data file
+        
+        Returns
+        -------
+        new_data_dict
+            A dictionary containing the data from data_dict
+            reformatted to be stored entirely as arrays, or as
+            one entry if the field does not change between records
+        """
+        
+        try:
+            new_data_dict = cls._iq_site_to_array(data_dict, 'bfiq')
```

Comment:
This is typically done inside the read classes before this point. I think it would be redundant here. If someone was doing something out of the ordinary it would be caught as a Restructure Error. I can also make this class not user-facing so that it would be difficult to use this code outside of the read/write classes

Reply:
sounds good to me.